src/sql/src/plan/statement/ddl.rs

```diff
@@ -4941,12 +4942,54 @@ pub fn plan_alter_object_rename(
         (ObjectType::ClusterReplica, UnresolvedObjectName::ClusterReplica(name)) => {
             plan_alter_cluster_replica_rename(scx, object_type, name, to_item_name, if_exists)
         }
+        (ObjectType::Schema, UnresolvedObjectName::Schema(name)) => {
+            plan_alter_schema_rename(scx, name, to_item_name, if_exists)
+        }
         (object_type, name) => {
             unreachable!("parser set the wrong object type '{object_type:?}' for name {name:?}")
         }
     }
 }
 
+pub fn plan_alter_schema_rename(
+    scx: &mut StatementContext,
+    name: UnresolvedSchemaName,
+    to_schema_name: Ident,
+    if_exists: bool,
+) -> Result<Plan, PlanError> {
+    scx.require_feature_flag(&vars::ENABLE_RENAME_SCHEMA)?;
+
+    let Some((db_spec, schema_spec)) = resolve_schema(scx, name.clone(), if_exists)? else {
+        let object_type = ObjectType::Schema;
+        scx.catalog.add_notice(PlanNotice::ObjectDoesNotExist {
+            name: name.to_ast_string(),
+            object_type,
+        });
+        return Ok(Plan::AlterNoop(AlterNoopPlan { object_type }));
+    };
+
+    // Make sure the name is unique.
+    if scx
+        .resolve_schema_in_database(&db_spec, &to_schema_name)
+        .is_ok()
+    {
+        return Err(PlanError::Catalog(CatalogError::SchemaAlreadyExists(
+            to_schema_name.clone().into_string(),
+        )));
+    }
+
+    // Prevent users from renaming system related schemas.
+    let schema = scx.catalog.get_schema(&db_spec, &schema_spec);
+    if schema.owner_id() == MZ_SYSTEM_ROLE_ID {
+        bail_never_supported!(format!("renaming the {} schema", schema.name().schema))
+    }
```

Comment:
I used `schema.owner_id()` because the "public" schema is a User schema, and I didn't want to allow folks to rename that, what do you think?

Reply:
The `public` schema in general is not owned by `mz_system`

```
materialize=> CREATE DATABASE db;
CREATE DATABASE
materialize=> SELECT * FROM mz_schemas WHERE name = 'public';
 id |  oid  | database_id |  name  | owner_id |            privileges             
----+-------+-------------+--------+----------+-----------------------------------
 u6 | 20603 | u2          | public | u1       | {s2=U/u1,u1=UC/u1,=U/u1}
 u3 | 20006 | u1          | public | s1       | {s1=UC/s1,s2=U/s1,u1=UC/s1,=U/s1}
(2 rows)

```
(Note the owner id of `u1`).

It's just that the `public` schema in the `materialize` database is owned by `mz_system` because there's no one else to own it. In general we treat `public` schemas the same as every other schema, it's just automatically created with some sane defaults. So I think we should allow users to rename the `public` schema.

Reply:
Makes sense! Updated to use `schema.id().is_system()`